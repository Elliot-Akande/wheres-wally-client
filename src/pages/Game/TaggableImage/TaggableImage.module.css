.imageContainer {
  width: 100%;
  margin-bottom: 96px;
  position: relative;

  border-radius: 16px;
}

.image {
  width: 100%;
  object-fit: contain;
  border-radius: 16px;
}

.placeholder {
  padding-top: 48px;
}

.marker {
  position: absolute;

  display: flex;
  align-items: center;
  justify-content: center;

  border-radius: 50px;
  transform: translate(-50%, -50%);
}

.correct {
  composes: marker;
  height: 25px;
  width: 25px;

  background-color: var(--accent-dark);
  box-shadow: 0 0 14px 14px var(--accent-dark);
  animation: fadein 0.3s;
}

.correctImg {
  max-height: 350%;
  max-width: 350%;
}

.incorrect {
  composes: marker;
  height: 16px;
  width: 16px;

  background-color: var(--error);
  box-shadow: 0 0 2px 2px var(--error);
  font-weight: bold;

  opacity: 0;
  animation: fadeout 1s cubic-bezier(0.075, 0.82, 0.165, 1);
}

@keyframes fadeout {
  25%,
  60% {
    transform: scale(2) translate(-25%, -25%);
    opacity: 1;
  }
}

@keyframes fadein {
  0% {
    transform: translate(-50%, -400%);
    opacity: 0;
  }

  /* 70% {
    transform: scale(1.1);
  } */
}
