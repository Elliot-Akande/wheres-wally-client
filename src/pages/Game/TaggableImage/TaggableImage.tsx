import { useRef, useState } from "react";
import IncorrectIcon from "../../../assets/incorrect.svg";
import ImageLoader from "../../../components/ImageLoader/ImageLoader";
import Magnifier from "../Magnifier/Maginifier";
import {
  CorrectContainer,
  CorrectImg,
  IncorrectImg,
  StyledTaggableImage,
  imageStyles,
} from "./styles";

interface Answer {
  character: string;
  xCoord: number;
  yCoord: number;
}

interface CorrectAnswer extends Answer {
  imageUrl: string;
}

interface TaggableImageProps {
  imageUrl: string;
  characters: Array<{
    name: string;
    imageUrl: string;
  }>;
  checkAnswer: (answer: Answer) => Promise<boolean>;
  correctAnswers: CorrectAnswer[];
  levelComplete: boolean;
}

type MagnifierBehaviour = "hidden" | "hover" | "clicked";

const TaggableImage = ({
  imageUrl,
  characters,
  checkAnswer,
  correctAnswers,
  levelComplete,
}: TaggableImageProps) => {
  const [clickedCoords, setClickedCoords] = useState({ x: 0, y: 0 });
  const [hoverCoords, setHoverCoords] = useState({ x: 0, y: 0 });
  const [magnifierBehaviour, setMagnifierBehaviour] =
    useState<MagnifierBehaviour>("hidden");
  const [showIncorrectMark, setShowIncorrectMark] = useState(false);
  const imageRef = useRef<HTMLImageElement>(null);

  const handleClick = (e: React.MouseEvent<HTMLDivElement>): void => {
    if (levelComplete) return;

    // Hide character selection menu
    if (magnifierBehaviour === "clicked") {
      setMagnifierBehaviour("hover");
      return;
    }

    const { top, left } = e.currentTarget.getBoundingClientRect();
    const x = Math.round(e.clientX - left);
    const y = Math.round(e.clientY - top);

    setShowIncorrectMark(false);
    setClickedCoords({ x, y });
    setMagnifierBehaviour("clicked");
  };

  const handleMouseMove = (e: React.MouseEvent<HTMLDivElement>): void => {
    const { top, left } = e.currentTarget.getBoundingClientRect();
    const x = Math.round(e.clientX - left);
    const y = Math.round(e.clientY - top);
    setHoverCoords({ x, y });
  };

  // Call checkAnswer prop with additional side effects
  const customCheckAnswer = async (character: string): Promise<void> => {
    setMagnifierBehaviour("hover");

    // Create answer object
    const img = imageRef.current;
    if (!img) return;

    const normalisedX = clickedCoords.x / img.width;
    const normalisedY = clickedCoords.y / img.height;
    const answer = {
      // Denormalised coords
      xCoord: Math.round(img.naturalWidth * normalisedX),
      yCoord: Math.round(img.naturalHeight * normalisedY),
      character,
    };

    if (!(await checkAnswer(answer))) {
      // Answer is incorrect
      setShowIncorrectMark(true);
      setTimeout(() => {
        setShowIncorrectMark(false);
      }, 1000);
    }
  };

  const getAnswerMarkerPositionStyles = (
    answer: Answer
  ): { left: string; top: string } => {
    // Normalise coords to current img dimensions
    const img = imageRef.current;
    if (!img)
      return {
        left: "0px",
        top: "0px",
      };

    const xCoord = answer.xCoord * (img.width / img.naturalWidth);
    const yCoord = answer.yCoord * (img.height / img.naturalHeight);

    return {
      left: `${xCoord}px`,
      top: `${yCoord}px`,
    };
  };

  const getImgDimensions = () => {
    if (imageRef.current === null) return { width: 0, height: 0 };

    const { width, height } = imageRef.current;
    return { width, height };
  };

  return (
    <StyledTaggableImage
      onClick={handleClick}
      onMouseMove={handleMouseMove}
      onMouseEnter={() => setMagnifierBehaviour("hover")}
      onMouseLeave={() => setMagnifierBehaviour("hidden")}
    >
      <ImageLoader
        src={imageUrl}
        alt="Where's Wally Game"
        ref={imageRef}
        imgStyles={imageStyles}
        spinnerColour="hsl(288, 70%, 49%)"
        placeholderStyles={{ paddingTop: "48px" }}
      />

      {/* Correct answer markers */}
      {correctAnswers.map((answer) => (
        <CorrectContainer
          key={answer.character}
          $positionStyles={getAnswerMarkerPositionStyles(answer)}
        >
          <CorrectImg src={answer.imageUrl} />
        </CorrectContainer>
      ))}

      <Magnifier
        clickedCoords={clickedCoords}
        hoverCoords={hoverCoords}
        magnifierBehaviour={magnifierBehaviour}
        characters={characters}
        checkAnswer={customCheckAnswer}
        imageDimensions={getImgDimensions()}
        imageUrl={imageUrl}
      />

      {showIncorrectMark && (
        <IncorrectImg
          src={IncorrectIcon}
          alt="Incorrect"
          $coords={clickedCoords}
        />
      )}
    </StyledTaggableImage>
  );
};

export default TaggableImage;
