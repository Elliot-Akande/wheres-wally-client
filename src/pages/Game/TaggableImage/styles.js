import { css, keyframes, styled } from "styled-components";

const fadeout = keyframes`
  25%,
  60% {
    transform: scale(2) translate(-25%, -25%);
    opacity: 1;
  }
`;

const fadein = keyframes`
  0% {
      transform: translate(-50%, -400%);
      opacity: 0;
    }
`;

export const StyledTaggableImage = styled.div`
  width: 100%;
  margin-bottom: 96px;
  position: relative;

  border-radius: 16px;
`;

const marker = css`
  position: absolute;

  display: flex;
  align-items: center;
  justify-content: center;

  border-radius: 50px;
  transform: translate(-50%, -50%);
`;

export const IncorrectImg = styled.img`
  ${marker}

  height: 16px;
  width: 16px;

  background-color: var(--error);
  box-shadow: 0 0 2px 2px var(--error);
  font-weight: bold;

  opacity: 0;
  animation: ${fadeout} 1s cubic-bezier(0.075, 0.82, 0.165, 1);

  ${({ $coords }) => css`
    left: ${$coords.x}px;
    top: ${$coords.y}px;
  `}
`;

export const CorrectContainer = styled.div`
  ${marker}

  height: 25px;
  width: 25px;

  background-color: var(--accent-dark);
  box-shadow: 0 0 14px 14px var(--accent-dark);
  animation: ${fadein} 0.3s;

  ${({ $positionStyles }) => {
    return $positionStyles;
  }}
`;

export const CorrectImg = styled.img`
  max-height: 350%;
  max-width: 350%;
`;
