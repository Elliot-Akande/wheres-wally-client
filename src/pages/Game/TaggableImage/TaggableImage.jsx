import { useState, useRef } from "react";
import Magnifier from "../Magnifier/Maginifier";
import styles from "./TaggableImage.module.css";
import ImageLoader from "../../../components/ImageLoader/ImageLoader";
import IncorrectIcon from "../../../assets/incorrect.svg";
import PropTypes from "prop-types";

const TaggableImage = ({
  imageUrl,
  characters,
  checkAnswer,
  correctAnswers,
  levelComplete,
}) => {
  const [clickedCoords, setClickedCoords] = useState({ x: 0, y: 0 });
  const [hoverCoords, setHoverCoords] = useState({ x: 0, y: 0 });
  const [magnifierBehaviour, setMagnifierBehaviour] = useState("hidden");
  const [showIncorrectMark, setShowIncorrectMark] = useState(false);
  const imageRef = useRef(null);

  const handleClick = (e) => {
    if (levelComplete) return;

    // Hide character selection menu
    if (magnifierBehaviour === "clicked") {
      setMagnifierBehaviour("hover");
      return;
    }

    const { top, left } = e.currentTarget.getBoundingClientRect();
    const x = Math.round(e.clientX - left);
    const y = Math.round(e.clientY - top);

    setShowIncorrectMark(false);
    setClickedCoords({ x, y });
    setMagnifierBehaviour("clicked");
  };

  const handleMouseMove = (e) => {
    const { top, left } = e.currentTarget.getBoundingClientRect();
    const x = Math.round(e.clientX - left);
    const y = Math.round(e.clientY - top);
    setHoverCoords({ x, y });
  };

  // Call checkAnswer prop with additional side effects
  const customCheckAnswer = async (character) => {
    setMagnifierBehaviour("hover");

    // Create answer object
    const img = imageRef.current;
    const normalisedX = clickedCoords.x / img.width;
    const normalisedY = clickedCoords.y / img.height;
    const answer = {
      // Denormalised coords
      xCoord: Math.round(img.naturalWidth * normalisedX),
      yCoord: Math.round(img.naturalHeight * normalisedY),
      character,
    };

    if (!(await checkAnswer(answer))) {
      // Answer is incorrect
      setShowIncorrectMark(true);
      setTimeout(() => {
        setShowIncorrectMark(false);
      }, 1000);
    }
  };

  const getAnswerMarkerPositionStyles = (data) => {
    // Normalise coords to current img dimensions
    const img = imageRef.current;
    const xCoord = data.xCoord * (img.width / img.naturalWidth);
    const yCoord = data.yCoord * (img.height / img.naturalHeight);

    return {
      left: `${xCoord}px`,
      top: `${yCoord}px`,
    };
  };

  const getImgDimensions = () => {
    if (imageRef.current === null) return { width: 0, height: 0 };

    const { width, height } = imageRef.current;
    return { width, height };
  };

  return (
    <div
      className={styles.imageContainer}
      onClick={handleClick}
      onMouseMove={handleMouseMove}
      onMouseEnter={() => setMagnifierBehaviour("hover")}
      onMouseLeave={() => setMagnifierBehaviour("hidden")}
    >
      <ImageLoader
        src={imageUrl}
        alt="Where's Wally Game"
        imgClass={styles.image}
        imgRef={imageRef}
        spinnerColour="hsl(288, 70%, 49%)"
        placeholderClass={styles.placeholder}
      />

      {/* Correct answer markers */}
      {correctAnswers.map((answer) => (
        <div
          key={answer.character}
          className={styles.correct}
          style={getAnswerMarkerPositionStyles(answer)}
        >
          <img src={answer.imageUrl} className={styles.correctImg} />
        </div>
      ))}

      <Magnifier
        clickedCoords={clickedCoords}
        hoverCoords={hoverCoords}
        magnifierBehaviour={magnifierBehaviour}
        characters={characters}
        checkAnswer={customCheckAnswer}
        imageDimensions={getImgDimensions()}
        imageUrl={imageUrl}
      />

      {showIncorrectMark && (
        <img
          src={IncorrectIcon}
          alt="Incorrect"
          className={styles.incorrect}
          style={{
            left: clickedCoords.x + "px",
            top: clickedCoords.y + "px",
          }}
        />
      )}
    </div>
  );
};

TaggableImage.propTypes = {
  imageUrl: PropTypes.string.isRequired,
  characters: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      imageUrl: PropTypes.string.isRequired,
    })
  ).isRequired,
  checkAnswer: PropTypes.func.isRequired,
  correctAnswers: PropTypes.arrayOf(
    PropTypes.shape({
      character: PropTypes.string.isRequired,
      imageUrl: PropTypes.string.isRequired,
      xCoord: PropTypes.number.isRequired,
      yCoord: PropTypes.number.isRequired,
    })
  ).isRequired,
  levelComplete: PropTypes.bool.isRequired,
};

export default TaggableImage;
