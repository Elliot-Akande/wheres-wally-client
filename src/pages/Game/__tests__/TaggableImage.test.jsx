import { describe, it, expect, vi } from "vitest";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import TaggableImage from "../TaggableImage/TaggableImage";

vi.mock("../Magnifier/Maginifier.jsx", () => ({
  default: ({ magnifierBehaviour }) => (
    <div data-testid="magnifier">{magnifierBehaviour}</div>
  ),
}));
vi.mock("../../../components/ImageLoader/ImageLoader.jsx", () => ({
  default: ({ src, alt }) => <img src={src} alt={alt} />,
}));
vi.mock("react", async (importOriginal) => {
  const mod = await importOriginal();
  return {
    ...mod,
    useRef: () => ({
      current: {
        width: 100,
        height: 100,
        naturalWidth: 500,
        naturalHeight: 500,
      },
    }),
  };
});

const mockProps = {
  imageUrl: "/test.jpg",
  characters: [
    { name: "John", imageUrl: "/john.jpg" },
    { name: "Kate", imageUrl: "/kate.jpg" },
  ],
  checkAnswer: vi.fn(),
  correctAnswers: [
    { character: "Kate", imageUrl: "/kate.jpg", xCoord: 10, yCoord: 10 },
  ],
  levelComplete: false,
};

describe("TaggableImage", () => {
  it("renders img, CorrectAnswers, and Magnifier", () => {
    const { container } = render(<TaggableImage {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it("sets magnifierBehaviour to 'hover' on mouseOver event", () => {
    render(<TaggableImage {...mockProps} />);

    const magnifier = screen.getByTestId("magnifier");
    expect(magnifier.textContent).toBe("hidden");

    const image = screen.getByRole("img", { name: "Where's Wally Game" });
    fireEvent.mouseOver(image);
    expect(magnifier.textContent).toBe("hover");
  });

  it("sets magnifierBehaviour to 'hidden' on mouseLeave event", () => {
    render(<TaggableImage {...mockProps} />);

    const magnifier = screen.getByTestId("magnifier");
    const image = screen.getByRole("img", { name: "Where's Wally Game" });
    fireEvent.mouseOver(image);
    expect(magnifier.textContent).toBe("hover");

    fireEvent.mouseLeave(image);
    expect(magnifier.textContent).toBe("hidden");
  });

  it("sets magnifierBehaviour to 'clicked' on click event", () => {
    render(<TaggableImage {...mockProps} />);

    const magnifier = screen.getByTestId("magnifier");
    expect(magnifier.textContent).toBe("hidden");

    const image = screen.getByRole("img", { name: "Where's Wally Game" });
    fireEvent.click(image);
    expect(magnifier.textContent).toBe("clicked");
  });

  it("sets magnifierBehaviour to 'hover' on click event when previously set to 'clicked'", () => {
    render(<TaggableImage {...mockProps} />);

    const magnifier = screen.getByTestId("magnifier");
    const image = screen.getByRole("img", { name: "Where's Wally Game" });
    fireEvent.click(image);
    expect(magnifier.textContent).toBe("clicked");

    fireEvent.click(image);
    expect(magnifier.textContent).toBe("hover");
  });

  it("does not handle click event when level is complete", () => {
    render(<TaggableImage {...mockProps} levelComplete={true} />);
    const image = screen.getByAltText("Where's Wally Game");
    fireEvent.click(image, { clientX: 10, clientY: 10 });
    expect(mockProps.checkAnswer).not.toHaveBeenCalled();
  });
});
