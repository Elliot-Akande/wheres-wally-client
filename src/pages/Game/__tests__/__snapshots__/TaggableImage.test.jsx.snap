// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaggableImage > renders img, CorrectAnswers, and Magnifier 1`] = `
<div>
  <div
    class="sc-beySPh iDmzAO"
  >
    <img
      alt="Where's Wally Game"
      src="/test.jpg"
    />
    <div
      class="sc-dmyCSP coJYrB"
    >
      <img
        class="sc-hLQSwg dChYok"
        src="/kate.jpg"
      />
    </div>
    <div
      data-testid="magnifier"
    >
      hidden
    </div>
  </div>
</div>
`;
